src/log.rs

```diff
@@ -0,0 +1,396 @@
+    match PRINT_HANDLER.lock() {
+        Ok(mut handler) => {
+            *handler = unsafe { ::std::mem::transmute(func) };
+        }
+        Err(_) => {
+            // should we log something?
+        }
+    }
+    unsafe { glib_sys::g_set_printerr_handler(callback) };
+}
+
+static PRINTERR_HANDLER: Lazy<Mutex<Option<Box_<Box_<dyn Fn(GString) + Send + Sync + 'static>>>>> =
+    Lazy::new(|| Mutex::new(None));
+
+pub fn set_printerr_handler<P: Fn(GString) + Send + Sync + 'static>(func: Option<P>) {
+    unsafe extern "C" fn func_func(string: *const libc::c_char) {
+        match PRINTERR_HANDLER.lock() {
+            Ok(handler) => {
+                if let Some(ref handler) = *handler {
+                    let string: GString = from_glib_borrow(string);
+                    (*handler)(string)
+                } else {
+                    panic!("PRINTERR_HANDLER cannot be None!");
+                }
+            }
+            Err(_) => {
+                // should we log something here?
+            }
+        }
+    }
+    let callback = if func.is_some() {
+        Some(func_func as _)
+    } else {
+        None
+    };
+    let func = func.map(|f| Box_::new(Box_::new(f)));
+    match PRINTERR_HANDLER.lock() {
+        Ok(mut handler) => {
+            *handler = unsafe { ::std::mem::transmute(func) };
+        }
+        Err(_) => {
+            // should we log something?
+        }
+    }
+    unsafe { glib_sys::g_set_printerr_handler(callback) };
+}
+
+pub fn log_set_default_handler<P: Fn(&str, &LogLevelFlags, &str) + Send + Sync + 'static>(
+    log_func: P,
+) {
+    let log_func_data: Box_<P> = Box_::new(log_func);
+    unsafe extern "C" fn log_func_func<
+        P: Fn(&str, &LogLevelFlags, &str) + Send + Sync + 'static,
+    >(
+        log_domain: *const libc::c_char,
+        log_level: glib_sys::GLogLevelFlags,
+        message: *const libc::c_char,
+        user_data: glib_sys::gpointer,
+    ) {
+        let log_domain: GString = from_glib_borrow(log_domain);
+        let log_level = from_glib(log_level);
+        let message: GString = from_glib_borrow(message);
+        let callback: &P = &*(user_data as *mut _);
+        (*callback)(log_domain.as_str(), &log_level, message.as_str());
+    }
+    let log_func = Some(log_func_func::<P> as _);
+    let super_callback: Box_<P> = log_func_data;
+    unsafe {
+        glib_sys::g_log_set_default_handler(log_func, Box_::into_raw(super_callback) as *mut _);
+    }
+}
+
+/// Macro used to log using GLib logging system. Is uses [g_log].
+///
+/// [g_log]: https://developer.gnome.org/glib/stable/glib-Message-Logging.html#g-log)
+///
+/// Example:
+///
+/// ```no_run
+/// use glib::{LogLevelFlags, g_log};
+///
+/// g_log!(None, LogLevelFlags::FLAG_RECURSION, "test");
+/// g_log!(Some("test"), LogLevelFlags::FLAG_FATAL, "test");
+///
+/// // You can also pass arguments like in format! or println!:
+/// let x = 12;
+/// g_log!(None, LogLevelFlags::FLAG_RECURSION, "test: {}", x);
```

Comment:
The flags should not be exposed as flags. They're an enum, with the flags part being internal. Check the docs :)

You basicall want `enum LogLevel { Error, Critical, Warning, Message, Info, Debug }`

Reply:
The problem is that we also have `FLAG_RECURSION` and `FLAG_FATAL`. I think this is why this is not an enum... So do we just ignore those two?

Reply:
They're marked as internal flags in the documentation, so yes :)

Reply:
I.e. ignore them and make this a proper enum

Reply:
Oh no:

```
Sets the log handler for a domain and a set of log levels. To handle fatal and recursive messages the log_levels parameter must be combined with the G_LOG_FLAG_FATAL and G_LOG_FLAG_RECURSION bit flags.
```

I'd handle these two flags as separate boolean parameters to `g_log_set_handler()`. The more tricky bit is `g_log_set_fatal_mask()`, which wants a mask of log levels.

But maybe we need both things then. For normal logging usage you really want an enum for simple and correct use.

Reply:
Oh indeed! Makes my life a lot easier. :) I'll keep the flags though (since we're having the C values generated, we should use them after all!), but we should doc hide them. Will check if it's already supported in gir, otherwise I'll add it.

Reply:
Ah. Then we can have both: a type builder for log levels which can be set as fatal?

Reply:
You must not pass `FATAL` or `RECURSIVE` to any function apart from `set_handler()`, and a combination of log levels (not the other flags) to `set_fatal_mask()`.

Reply:
Ok, so I need to create two separate enums.

Reply:
One enum and one or two flags types, I guess. One flags type if you handle `FATAL` / `RECURSIVE` as additional booleans in the API.